src/gfx/shaders/Volume_frag.glsl

```diff
@@ -42,85 +45,90 @@ vec4 sample3DTexture(vec3 texCoord)
   return mix(colorSlice1, colorSlice2, weightSlice2);
 }
 
-vec4 sample3DTextureInclined(vec3 boxCoord) { //delta:{ x: XY, y : XZ, z: YZ }
+vec4 sample3DTextureInclined(vec3 boxCoord) { // delta:{ x: XY, y : XZ, z: YZ }
   vec3 textCoord = boxCoord;
   vec2 currDelta = mix(boxCoord.zz, vec2(1., 1.) - boxCoord.zz, boxAngles.yx) * delta.yz;
 
   textCoord.y = (boxCoord.y  - currDelta.y) / (1. - delta.z);
   if (textCoord.y < 0.0 || textCoord.y > 1.0)
-    return vec4(0., 0., 0., 0);
+    return noColor;
 
   currDelta.x += mix(textCoord.y, 1.0 - textCoord.y, boxAngles.z) * delta.x;
 
   textCoord.x = (boxCoord.x - currDelta.x) / (1. - delta.x - delta.y);
   if (textCoord.x < 0.0 || textCoord.x > 1.0)
-    return vec4(0., 0., 0., 0);
+    return noColor;
 
   return sample3DTexture(textCoord);
 }
 
-float CalcColor(vec3 iter, vec3 dir)
-{
-  float d = 1.0 / 128.0;
+float CalcColor(vec3 iter, vec3 dir) {
+  float d = 1. / 128.;
   vec3 dx = vec3(d, 0.0, 0.0);
   vec3 dy = vec3(0.0, d, 0.0);
   vec3 dz = vec3(0.0, 0.0, d);
-  vec3 N;
-  N.x = sample3DTextureInclined(iter + dx).r - sample3DTextureInclined(iter - dx).r;
-  N.y = sample3DTextureInclined(iter + dy).r - sample3DTextureInclined(iter - dy).r;
-  N.z = sample3DTextureInclined(iter + dz).r - sample3DTextureInclined(iter - dz).r;
-  N = normalize(N);
-  float dif = max(0.0, dot(N,dir));
+
+  vec3 coordInc = vec3(
+    // #Opt: coordInc.x:(iter + dx).x > 1. ? 0.: sample3DTextureInclined(iter + dx).x,
+    mix(sample3DTextureInclined(iter + dx).x, 0., floor((iter + dx).x)),
```

Comment:
Please forgive me this recurring question but isn't it possible to vectorize the calculations? I'm also a little bit concerned with the growth of calculations while we need to reduce/optimize them.

Reply:
I rewrote this part of the code so now it theoretically should work not worse than before, but in fact, even under heavy load, I couldn't see the difference.